### PR TITLE
Fixing when keyword recommender in recursive patterns branch #2

### LINKED ISSUE #30848: When keyword recommender failure with var patterns
A [recent change to the syntax](https://github.com/dotnet/roslyn/issues/29297) to support `var` patterns caused `WhenKeywordRecommenderTests` to fail in the `features/recursive-patterns` branch (during CI tests). They are skipped with this bug ID.

/cc @Neme12

### PR DESCRIPTION
**Pull request by Neme12**
Fixes #30848

cc @gafter

### CODE REVIEW

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 1]
```diff
@@ -455,13 +482,43 @@ class C
...
         public async Task TestForSwitchCase_SemanticCheck_AfterClassAndLocalConstantVar_BeforeWhen() =>
     await VerifyKeywordAsync(@"
 class var { }
 class C
 {
     void M() { const object var = null; switch (new object()) { case var $$ when } }
 }");
+
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        public async Task TestForSwitchCase_SemanticCheck_AfterTypeAliasAndFieldConstantVar() =>
```
Comment (Neme12) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: This is not new behavior, it worked before the tests were skipped. Just adding an additional test case.

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 2]
```diff
@@ -425,19 +425,46 @@ class C
     void M() { switch (new object()) { case var $$ when } }
 }");
 
-        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/30848"), Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
```
Comment (Neme12) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: This is not new behavior, it worked before the tests were skipped. Just adding an additional test case.

**Comment by Neme12** [comment]
@dpoeschl Would you please review this?

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 3]
```diff
@@ -425,19 +425,46 @@ class C
...
+        public async Task TestForSwitchCase_SemanticCheck_NotAfterTypeAliasVar() =>
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: 1. methods with => have the=> on hte next line in IDE code.
2. IDE test methods don't use =>, they use normal blocks. 

Could you follow the patterns of the rest of the IDE code here?  We can certainly consider a PR in the future to update tests to be of this form if people liek it.

**Comment by CyrusNajmabadi** [comment]
PR easier to review as: https://github.com/dotnet/roslyn/pull/30870/files?w=1 due to indentation changes.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs [hunk 4]
```diff
@@ -108,22 +124,38 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
             // If what we have here is known to be a type, we don't want to clutter the variable name suggestion list
             // with 'when' since we know that the resulting code would be semantically invalid.
 
-            var expression = expressionOrPattern as ExpressionSyntax
-                ?? (expressionOrPattern as ConstantPatternSyntax)?.Expression;
-            
-            if (!(expression is TypeSyntax typeSyntax))
+            bool isVar;
+            ImmutableArray<ISymbol> symbols;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs]: i think it woudl be worth documenting these.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs [hunk 5]
```diff
@@ -132,9 +164,9 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
                 // assumption that the user didn't just type 'var' to then create a constant named 'var', but really
                 // is about to declare a variable. Therefore we don't want to interfere with the declaration.
                 // However note that if such a constant already exists, we do the right thing and do offer 'when'.
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs]: great comment.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs [hunk 6]
```diff
@@ -108,22 +124,38 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
...
             // with 'when' since we know that the resulting code would be semantically invalid.
 
-            var expression = expressionOrPattern as ExpressionSyntax
-                ?? (expressionOrPattern as ConstantPatternSyntax)?.Expression;
-            
-            if (!(expression is TypeSyntax typeSyntax))
+            bool isVar;
+            ImmutableArray<ISymbol> symbols;
+
+            if (nodeOrToken.IsNode)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs]: i think it woudl be worth documenting which cases each part of the if is trying to handle.

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs (hunk 3 above)
Comment (Neme12) [reply src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: ➡️ Fixed in 7b1c2a6028ff3423790e02d7001a87ed52e349d7

**Comment by Neme12** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by gafter** [comment]
@Neme12 Thank you!!!!!

**Pull request merged**